File: exercises/D805/solution.cpp
#include <iostream>
using namespace std;
int a[101], b[101];
int main() {
    int n, m;
    cin >> n >> m;
    for (int i = 1; i <= n; i++) {
        cin >> a[i];
    }
    int l1, r1, l2, r2;
    for (int i = 1; i <= m; i++) {
        cin >> l1 >> r1 >> l2 >> r2;
        int j = l1, k = l2, t = l1;
        while (t <= r2) {
            // short circuit: k == m is checked first
            if (k > r2 || j <= r1 && a[j] < a[k]) {
                b[t++] = a[j++];
            } else {
                b[t++] = a[k++];
            }
        }
        for (j = l1; j <= r2; j++) {
            a[j] = b[j];
            cout << a[j] << (j == r2 ? '\n' : ' ');
        }
    }
}

File: exercises/D805/solution.py
n, m = map(int, input().split())
a = list(map(int, input().split()))
a = [0] + a
b = [0 for _ in range(n + 1)]
for i in range(m):
    l1, r1, l2, r2 = map(int, input().split())
    j = l1
    k = l2
    t = l1
    while t <= r2:
        # short circuit: k == m is checked first
        if k > r2 or j <= r1 and a[j] < a[k]:
            b[t] = a[j]
            j += 1
        else:
            b[t] = a[k]
            k += 1
        t += 1
    for j in range(l1, r2 + 1):
        a[j] = b[j]
        print(a[j], end='')
        if j == r2: 
            print() 
        else:
            print(end=' ')